dbr quant: record dag of non-quantizeable ops

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71781 dbr quant function fusion [2/x]: use fusion for observation and inference
* #71780 fx quant: add fusion matching for operator.add and torch.relu
* #71764 dbr quant function fusion [1/x]: record matches for functions
* **#71551 dbr quant: record dag of non-quantizeable ops**
* #71324 dbr quant: refactor first_call hooks to be standalone
* #71312 dbr quant: rename seen_op_info to seen_q_op_info

Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33684130](https://our.internmc.facebook.com/intern/diff/D33684130)

vkuzo committed:
dbr quant: record dag of non-quantizeable ops

Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant: record dag of non-quantizeable ops"


Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33684130).

vkuzo committed:
Update on "dbr quant: record dag of non-quantizeable ops"


Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33684130](https://our.internmc.facebook.com/intern/diff/D33684130)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33684130).

jerryzh168 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -114,36 +117,51 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             if hook_type is HookType.OP_HOOKS:
-                qstate = parent_module._auto_quant_state  # type: ignore[attr-defined]
                 fqn = module_id_to_fqn[id(parent_module)] if parent_module else None
+                qstate = parent_module._auto_quant_state  # type: ignore[attr-defined]
```
are we coupling tracing with quantization logic here?

vkuzo replied:
oh, sorry, could you explain more? not sure what you mean yet

jerryzh168 replied:
oh sorry, I mean we are querying auto_qaunt_state here in auto_trace.py, which means auto_trace is coupled with quantization and can't be applied to other transformations, e.g. numeric suite?

vkuzo replied:
oh, I see.  Yes, currently it is coupled.  There could be ways to reduce the coupling, but it's not something we have planned for this half.

jerryzh168 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -114,36 +117,51 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
                     output = qstate.op_prepare_after_hook(
                         func, output, args, global_op_idx)
                 qstate.mark_cur_op_complete(func)
             else:
```
nit: can we add a comment for what this `else` matches to and what this branch means?

jerryzh168 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -114,36 +117,51 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+                    if qstate:
+                        fqn = module_id_to_fqn.get(id(parent_module), None) \
+                            if parent_module else None
+                        is_quantizeable_op = False
```
can we add a comment on when this is True and when this is False? not sure I understand from the code

vkuzo replied:
sure, I can clean this op, maybe this is better expressed as an enum

jerryzh168 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -114,36 +117,51 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+
                 output = super().__torch_function__(func, types, args, kwargs)
 
+                if first_call:
```
we have a lot of `if first call`, wondering if we can separate all first call logic into one function?

vkuzo replied:
yeah, I've thought about this too.  Maybe we can consider it in the future as a refactor-only PR? lmk your thoughts

jerryzh168 replied:
yeah that sounds good

jerryzh168 commented on torch/ao/quantization/_dbr/quantization_state.py:
```diff
@@ -77,8 +78,15 @@ def __init__(
+        # unquantizeable ops are recorded without an index, because
+        # we do not enforce order of execution.
         self.idx_to_seen_q_op_infos: Dict[int, SeenQOpInfo] = {}
+        self.seen_nonq_op_infos: List[SeenNonQOpInfo] = []
```
why do we record non-quantizable in a separate list? do we need order between quantizable and non-quantizable ops?

can we also add a comment for what the seen_nonq_op_infos are used for?

vkuzo replied:
yeah, this one is a tricky one.  There are two reasons:
1. for quantizeable ops, we need to enforce that ops seen during non-first-call are in the same order as ops seen during tracing, or throw an exception
2. for quantziable ops, the dynamic logic is called at every forward, so it needs to be optimized for performance

Currently the code is separated because of the two things above, for quantizeable ops the restrictions are different and performance matters, at the expense of readability.  There might be a way to reorganize it, but it would have to not regress performance.  Probably something to consider in the future, although I'm not sure of a clean way to do that at the moment, lmk if you have thoughts.

jerryzh168 replied:
I see, one thing for performance is we can explore the path to construct an fx graph I think, we can discuss later

vkuzo replied:
makes sense, for the FX graph, it would be helping with performance of executing subgraphs.  One thing which will still need to be done without FX is the performance of matching quantizeable ops seen during tracing to quantizeable ops seen afterwards.  For example,

```
def forward(self, x):
  x = x + x
  x = x * x
  return x
```

even if we have an FX graph for add -> mul, we'd need to match that the user executes an add and then a mul to verify that there is no data dependent control flow.  Note: this may be relaxed with some kind of "unsafe_do_not_check_seen_ops" flag, which would have to be off by default.

jerryzh168 replied:
you mean during execution we need to check that the data dependency does not change? Is this the same as the "validating trace" step mentioned in Ed's doc?

vkuzo replied:
Yes.  It could potentialy be disabled, but that's something the user would disable themselves because they would be skipping safety checks, by default we need to validate the assumption every time, since we do not capture the entire program.

vkuzo replied:
also, if there is some other user code that is not captured (non-pytoch ops), we would still need to execute it.

jerryzh168 replied:
makes sense, I guess we will need to have a speculative execution framework for that, as mentioned in the doc, to mitigate the impact on performance, for now I think we can add a comment briefly explaining the reason that we have separate lists, and purpose for each list. we can follow up later on the feasibility of optimizing them

vkuzo replied:
I think adding things like speculative execution framework would also significantly (2x or more) increase the technical complexity of the project, I'd probably want to verify that something like that is actually needed by real use cases before investing in it.

jerryzh168 replied:
yeah we probably need support from other teams for this

vkuzo replied:
Oh, I mean that IMO adding speculative execution to this project might not make sense at all, whether we have help from other teams or not.  It's not going to be free, it will make things slower and harder to maintain.  At some point, we have to consider the technical complexity of the project versus the engineering cost, maintenance cost, value to users.

vkuzo replied:
Here is one way to think about it:
1. how do we define general quantziation: for all possible paths throughout the program (arbitrary control flow), or for a single linear path
2. how do we implement whatever we decided to define in (1)

Currently, we are defining (1) to only work for a linear path of quantizeable ops which does not change.  There is no need for a specalutive execution framework to implement that.

If we wanted to define (1) more generally (we decided not to in the design review), speculative execution would be one way to implement a more general definition of quantization, assuming that we don't have a fully captured program.

Basically, if we are doing it the simple way, there shouldn't be any need for speculative execution anyways.

> makes sense, I guess we will need to have a speculative execution framework for that, as mentioned in the doc, to mitigate the impact on performance

One thing to note is that we are still running every line of the user code, because there is no full program capture.  Even if there is a subgraph behind the scenes, there is still a performance cost to run the user code and check that it matches what we expect.  That cost would still be there with speculative execution, it would not be amortized away.

jerryzh168 replied:
let's follow up in meetings

vkuzo committed:
Update on "dbr quant: record dag of non-quantizeable ops"


Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33684130](https://our.internmc.facebook.com/intern/diff/D33684130)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33684130).

jerryzh168 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -38,9 +39,12 @@ def add_auto_observation(
 ) -> torch.nn.Module:
+    if prepare_custom_config_dict is None:
+        prepare_custom_config_dict = {}
+    output_dtypes = prepare_custom_config_dict.get('output_dtypes', (torch.float,))
```
is this a refactor? not sure where is this used

vkuzo replied:
this is to unbreak the `test_inplace_add` case.  I believe the FX comparison currently has the FX graph mode model return a long dtype (which might be a bug in FX graph mode quantization).  This code just makes sure the test keeps passing.

vkuzo committed:
Update on "dbr quant: record dag of non-quantizeable ops"


Summary:

This PR makes DBR quant record the DAG of non-quantizeable
ops. Having this will enable us to analyze the entire traced
graph of pytorch ops, regardless of whether they support quantization
or not.  That, in turn, will enable analysis of uses of each op,
allowing us to safely determine whether a subgraph of ops can be
fused or not.

In future PRs, this functionality will be used to implement function
fusion.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33684130](https://our.internmc.facebook.com/intern/diff/D33684130)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33684130).
